### PR TITLE
Checked `StrictMVar`s and `StrictTVar`s

### PR DESCRIPTION
**Pull request by jorisdral**
This PR moves checked versions of `StrictMVar` and `StrictTVar` to `cardano-base`.

Related PRs:
* https://github.com/input-output-hk/io-sim/pull/108
* https://github.com/input-output-hk/io-sim/pull/109

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -41,3 +42,12 @@ if impl(ghc >= 9.6)
...
     , protolude:text
+
+-- TODO: remove when a new version of strict-mvar (>1.1.0.0) is released
+source-repository-package
+  type: git
+  location: https://github.com/input-output-hk/io-sim
+  tag: 1b2c22b376f5cda314b9ab444caaf77764961a18
+  --sha256: 1qdxcqr4sl93n036p3pz1rvk6zf4qbsadzw4lgsp9agkh8pvs16y
+  subdir:
+    strict-mvar
```
Comment (jorisdral) [inline cabal.project:53]: `strict-mvar-1.1.0.0` doesn't export `LazyMVar` yet, but it should in a next version

**Comment by lehins** [comment]
What is the plan for`strict-checked-vars`, will it be released to hackage or CHaP?

File: strict-checked-vars/src/Control/Concurrent/Class/MonadMVar/Strict/Checked.hs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      -- Returning @a'@ along with @b@ allows us to check the invariant /after/
+      -- filling in the MVar.
+      pure (a' , (a', b))
+
+-- | 'modifyMVarMasked_' is defined in terms of 'modifyMVarMasked'.
+modifyMVarMasked_ :: (HasCallStack, MonadMVar m)
+                  => StrictMVar m a
+                  -> (a -> m a)
+                  -> m ()
+modifyMVarMasked_ v io = modifyMVar v io'
```
Comment (coot) [inline strict-checked-vars/src/Control/Concurrent/Class/MonadMVar/Strict/Checked.hs:164]:
```suggestion
modifyMVarMasked_ v io = modifyMVarMasked v io'
```

File: strict-checked-vars/strict-checked-vars.cabal [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+extra-doc-files:
+  CHANGELOG.md
+  README.md
+
+bug-reports:     https://github.com/input-output-hk/io-sim/issues
+tested-with:     GHC ==8.10 || ==9.2 || ==9.4 || ==9.6
+
+source-repository head
+  type:     git
+  location: https://github.com/input-output-hk/ouroboros-consensus
```
Comment (coot) [inline strict-checked-vars/strict-checked-vars.cabal:31]:
```suggestion
  location: https://github.com/input-output-hk/cardano-base
```

File: strict-checked-vars/strict-checked-vars.cabal [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+copyright:       2019-2023 Input Output Global Inc (IOG).
+author:          IOG Engineering Team
+maintainer:      operations@iohk.io
+category:        Concurrency
+build-type:      Simple
+extra-doc-files:
+  CHANGELOG.md
+  README.md
+
+bug-reports:     https://github.com/input-output-hk/io-sim/issues
```
Comment (coot) [inline strict-checked-vars/strict-checked-vars.cabal:26]:
```suggestion
bug-reports:     https://github.com/input-output-hk/cardano-base/issues
```

File: strict-checked-vars/strict-checked-vars.cabal [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  build-depends:
+    , base         >=4.9  && <4.19
+    , io-classes   ^>=1.1
+    , strict-mvar  ^>=1.1
+    , strict-stm   ^>=1.1
+
+  ghc-options:
+    -Wall -Wno-unticked-promoted-constructors -Wcompat
+    -Wincomplete-uni-patterns -Wincomplete-record-updates
+    -Wpartial-fields -Widentities
```
Comment (coot) [inline strict-checked-vars/strict-checked-vars.cabal:50]: I suspect that `-Wno-unticked-promoted-constructors` is not really needed.  In `io-classes` we use these flags:
```
    ghc-options: -Wall
                 -Wcompat
                 -Wincomplete-uni-patterns
                 -Wincomplete-record-updates
                 -Wpartial-fields
                 -Widentities
                 -Wunused-packages
                 -Wno-redundant-constraints
                 -Wno-unticked-promoted-constructors
```

File: strict-checked-vars/strict-checked-vars.cabal [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    , strict-checked-vars
+    , strict-mvar
+    , strict-stm
+    , tasty
+    , tasty-quickcheck
+
+  ghc-options:
+    -Wall -Wno-unticked-promoted-constructors -Wcompat
+    -Wincomplete-uni-patterns -Wincomplete-record-updates
+    -Wpartial-fields -Widentities -fno-ignore-asserts
```
Comment (coot) [inline strict-checked-vars/strict-checked-vars.cabal:75]: as above

**Review by coot — CHANGES_REQUESTED** [review]
Thanks @jorisdral.

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**